File: src/print_partner/core/import_rules.py
"""Project-level STL import rules (opt-in file/folder selection)."""

from __future__ import annotations

import json
from pathlib import Path


def normalize_relative_path(path: str) -> str:
    return path.replace("\\", "/").strip().lstrip("/")


def normalize_rule(rule: str) -> str:
    """Normalize a single import rule (file path or folder prefix with trailing /)."""
    r = normalize_relative_path(rule)
    if not r:
        return r
    if r.endswith("/"):
        return r
    # Treat as folder prefix if it has no .stl extension
    if r.lower().endswith(".stl"):
        return r
    return f"{r}/"


def parse_import_rules_json(raw: str | None) -> list[str] | None:
    """
    Parse Project.imported_paths column.
    Returns None for legacy (import all), else a list of rules (may be empty).
    """
    if raw is None:
        return None
    raw = raw.strip()
    if not raw:
        return []
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return []
    if not isinstance(data, list):
        return []
    rules: list[str] = []
    for item in data:
        if isinstance(item, str) and item.strip():
            rules.append(normalize_rule(item))
    return rules


def serialize_import_rules(rules: list[str] | None) -> str | None:
    """Serialize rules for DB. None = legacy all; [] = opt-in none."""
    if rules is None:
        return None
    normalized = []
    seen: set[str] = set()
    for rule in rules:
        r = normalize_rule(rule)
        if r and r not in seen:
            seen.add(r)
            normalized.append(r)
    return json.dumps(normalized)


def path_matches_rules(relative_path: str, rules: list[str]) -> bool:
    """True if relative_path is included by any rule."""
    if not rules:
        return False
    norm = normalize_relative_path(relative_path)
    for rule in rules:
        if rule.endswith("/"):
            prefix = rule.rstrip("/")
            if norm == prefix or norm.startswith(prefix + "/"):
                return True
        elif norm == rule:
            return True
    return False


def list_stl_relative_paths(repo_root: Path) -> list[str]:
    """List all STL paths relative to repo_root (sorted)."""
    if not repo_root.is_dir():
        return []
    root = repo_root.resolve()
    paths: list[str] = []
    for stl in root.rglob("*.stl"):
        if stl.is_file():
            paths.append(normalize_relative_path(str(stl.relative_to(root))))
    return sorted(paths, key=str.lower)


def count_matching_stls(repo_root: Path, rules: list[str] | None) -> tuple[int, int]:
    """Return (matching, total) STL counts under repo_root."""
    all_paths = list_stl_relative_paths(repo_root)
    total = len(all_paths)
    if rules is None:
        return total, total
    matching = sum(1 for p in all_paths if path_matches_rules(p, rules))
    return matching, total


def rules_from_selected_files(selected_files: list[str]) -> list[str]:
    """
    Compress a list of selected file paths into minimal rules (files + folder prefixes).
    Caller may pass explicit file paths; we store files individually and folder prefixes
    when an entire directory is selected (handled by UI).
    """
    rules: list[str] = []
    seen: set[str] = set()
    for f in selected_files:
        r = normalize_rule(f) if f.endswith("/") else normalize_relative_path(f)
        if r and r not in seen:
            seen.add(r)
            rules.append(r)
    return rules


def import_rules_for_project(imported_paths_raw: str | None) -> list[str] | None:
    """Read rules from Project.imported_paths column."""
    return parse_import_rules_json(imported_paths_raw)


def expand_rules_to_files(repo_root: Path, rules: list[str]) -> list[str]:
    """Expand rules to explicit file paths under repo_root."""
    all_paths = list_stl_relative_paths(repo_root)
    return [p for p in all_paths if path_matches_rules(p, rules)]

File: web/packages/domain/src/import-rules.ts
/** Project-level STL import rules (ported from Python import_rules.py). */

export function normalizeRelativePath(path: string): string {
  return path.replace(/\\/g, "/").trim().replace(/^\/+/, "");
}

export function normalizeRule(rule: string): string {
  const r = normalizeRelativePath(rule);
  if (!r) return r;
  if (r.endsWith("/")) return r;
  if (r.toLowerCase().endsWith(".stl")) return r;
  return `${r}/`;
}

export function parseImportRulesJson(raw: string | null | undefined): string[] | null {
  if (raw == null) return null;
  const trimmed = raw.trim();
  if (!trimmed) return [];
  try {
    const data = JSON.parse(trimmed) as unknown;
    if (!Array.isArray(data)) return [];
    const rules: string[] = [];
    for (const item of data) {
      if (typeof item === "string" && item.trim()) {
        rules.push(normalizeRule(item));
      }
    }
    return rules;
  } catch {
    return [];
  }
}

export function serializeImportRules(rules: string[] | null): string | null {
  if (rules == null) return null;
  const normalized: string[] = [];
  const seen = new Set<string>();
  for (const rule of rules) {
    const r = normalizeRule(rule);
    if (r && !seen.has(r)) {
      seen.add(r);
      normalized.push(r);
    }
  }
  return JSON.stringify(normalized);
}

export function pathMatchesRules(relativePath: string, rules: string[]): boolean {
  if (!rules.length) return false;
  const norm = normalizeRelativePath(relativePath);
  for (const rule of rules) {
    if (rule.endsWith("/")) {
      const prefix = rule.slice(0, -1);
      if (norm === prefix || norm.startsWith(`${prefix}/`)) return true;
    } else if (norm === rule) {
      return true;
    }
  }
  return false;
}

export function compressRulesFromFiles(checkedFiles: Iterable<string>): string[] {
  const allFiles = new Set(checkedFiles);
  if (allFiles.size === 0) return [];

  const dirPrefixes = new Set<string>();
  for (const f of allFiles) {
    const parts = f.split("/");
    for (let i = 1; i < parts.length; i++) {
      dirPrefixes.add(parts.slice(0, i).join("/"));
    }
  }

  const rules: string[] = [];
  const used = new Set<string>();

  for (const prefix of [...dirPrefixes].sort(
    (a, b) => b.split("/").length - a.split("/").length || a.localeCompare(b),
  )) {
    const prefixFiles = [...allFiles].filter((f) => f.startsWith(`${prefix}/`));
    if (!prefixFiles.length) continue;
    if (prefixFiles.every((f) => used.has(f))) continue;
    if (prefixFiles.every((f) => allFiles.has(f))) {
      rules.push(normalizeRule(`${prefix}/`));
      for (const f of prefixFiles) used.add(f);
    }
  }

  for (const f of [...allFiles].sort()) {
    if (!used.has(f)) rules.push(f);
  }
  return rules;
}

export function importRulesForProject(importedPathsRaw: string | null | undefined): string[] | null {
  return parseImportRulesJson(importedPathsRaw ?? null);
}
